src/client/dfuse/ops/setxattr.c

```diff
@@ -14,21 +14,44 @@ dfuse_cb_setxattr(fuse_req_t req, struct dfuse_inode_entry *inode,
 		  const char *name, const char *value, size_t size,
 		  int flags)
 {
-	int rc;
+	int	rc;
+	bool	duns_attr = false;
 
 	DFUSE_TRA_DEBUG(inode, "Attribute '%s'", name);
 
 	if (strcmp(name, DUNS_XATTR_NAME) == 0) {
 		struct duns_attr_t	dattr = {};
 
+		/* Just check this is valid, but don't do anything with it */
 		rc = duns_parse_attr((char *)value, size, &dattr);
 		if (rc)
 			D_GOTO(err, rc);
+
+		duns_attr = true;
 	}
 
 	rc = dfs_setxattr(inode->ie_dfs->dfs_ns, inode->ie_obj, name, value,
 			  size, flags);
 	if (rc == 0) {
+		/* Optionally remove the dentry to force a new lookup on access.
+		 * If the xattr is to set a UNS entry point, and dentry_dir
+		 * caching is enabled then invalidate the dentry here, to force
+		 * a lookup which will check the xattr and return the linked
+		 * container.  The fuse header says this potentially deadlocks
+		 * however it does appear to work, and calling this after the
```

Comment:
should add a test for this

Reply:
test_uns_create_with_cache() covers this, without the hook here it creates the UNS, then when you stat it you get the underlying file as it comes from the dentry cache, with this hook it works as expected, so if we hit the theoretical race condition then the test would deadlock.

It's not ideal as I put in the comment and I'm intending to raise this on the fuse mailing list for clarification.